keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -0,0 +1,1182 @@
+			{Version: mustParseVersion("1.27"), Default: false, PreRelease: Alpha},
+		},
+    featureB: VersionedSpecs{
+			{Version: mustParseVersion("1.28"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("1.30"), Default: true, PreRelease: Beta},
+		},
+}
+
+type featureOptions struct {
+  AEnabled        bool
+  AHasCapabilityZ bool
+  BEnabled        bool
+  BHandler        func()
+}
+
+func newFeatureOptions(featureGate FeatureGate) featureOptions {
+  opts := featureOptions{}
+  if featureGate.Enabled(FeatureA) {
+    opts.AEnabled = true
+  }
+  if featureGate.CompatibilityVersion() > "1.29" {
+    opts.AHasCapabilityZ = true
+  }
+
+  if featureGate.Enabled(FeatureB) {
+    opts.BEnabled = true
+  }
+  if featureGate.CompatibilityVersion() > "1.28" {
+    opts.BHandler = newHandler
+  } else {
+    opts.BHandler = oldHandler
+  }
+  return opts
+}
+
+// in client.go
+func ClientFunction() {
+  // ...
+  featureOpts := newFeatureOptions(utilfeature.DefaultFeatureGate)
+  if featureOpts.AEnabled {
+    // ...
+    if featureOpts.AHasCapabilityZ {
+      // run CapabilityZ
+    }
+  }
+  if featureOpts.BEnabled {
+    featureOpts.BHandler()
+  }
+  // ...
+}
+
+```
+
+### CEL Environment Compatibility Versioning
+
+CEL environments already [support a compatibility
+version](https://github.com/kubernetes/kubernetes/blob/7fe31be11fbe9b44af262d5f5cffb1e73648aa96/staging/src/k8s.io/apiserver/pkg/cel/environment/base.go#L45).
+The CEL compatibility version is used to ensure when a kubeneretes contol plane
+component reads a CEL expression from storage written by a (N+1) newer version
+(either due to version skew or a rollback), that a compatible CEL environment
+can still be constructed and the expression can still be evaluated.  This is
+achieved by making any CEL environment changes (language settings, libraries,
+variables) available for [stored expressions one version before they are allowed
+to be written by new
+expressions](https://github.com/kubernetes/kubernetes/blob/7fe31be11fbe9b44af262d5f5cffb1e73648aa96/staging/src/k8s.io/apiserver/pkg/cel/environment/environment.go#L38).
+
+The only change we must make for this enhancement is to remove the
+[compatibility version
+constant](https://github.com/kubernetes/kubernetes/blob/7fe31be11fbe9b44af262d5f5cffb1e73648aa96/staging/src/k8s.io/apiserver/pkg/cel/environment/base.go#L45)
+and instead always pass in N-1 of the compatibility version introduced by this
+enhancement as the CEL compatibility version.
+
+### StorageVersion Compatibility Versioning
+
+StorageVersions specify what version an apiserver uses to write resources to etcd
+for each API group. The StorageVersion changes across releases as API groups
+graduate through stability levels.
+
+During upgrades and downgrades, the storage version is particularly important.
+To enable upgrades and rollbacks, the version selected for storage in etcd in 
+version N must be (en/de)codable for k8s versions N-1 through N+1.
```

Comment:
Does this exclude alpha storage versions? On the promotion from alpha to beta, I don't want a new server serving a beta API for the first time to be forced to choose to store in etcd in alpha format, merely because the previous version understood it.

Reply:
> On the promotion from alpha to beta, I don't want a new server serving a beta API for the first time to be forced to choose to store in etcd in alpha format, merely because the previous version understood it.

This also came up earlier when discussion validatingadmissionpolicy's upgrade from alpha to beta. 

https://github.com/kubernetes/kubernetes/pull/118644#discussion_r1258879310
https://github.com/kubernetes/kubernetes/pull/120018#issuecomment-1682969549

I'm not sure from that I have a firm conclusion on what is best to do here. If we are forcing new beta apis to use the beta storage version even though the N-1 version does not understand it, doesn't this break rollbacks if the feature is enabled?